Test failures with PyPy3.9/3.10 7.3.16

### What pipdeptree version are you using?

2.21.0

### Are you running pipdeptree in a virtual environment?

No

### Describe the problem

The test suite fails both in PyPy3.9 and PyPy3.10, as of version 7.3.16. FWICS, #302 attempted to address this but I don't see a tracker bug for the problems.

I'm including the output both for PyPy3.10. PyPy3.9 fails in exactly the same way.

```pytb
$ tox -e pypy310
.pkg: _optional_hooks> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: get_requires_for_build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: get_requires_for_build_editable> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: build_wheel> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True hatchling.build
pypy310: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/pipdeptree/.tox/.tmp/package/4/pipdeptree-2.21.0-py3-none-any.whl
pypy310: commands[0]> python -m pytest --cov /tmp/pipdeptree/.tox/pypy310/lib/pypy3.10/site-packages/pipdeptree --cov /tmp/pipdeptree/tests --cov-config=pyproject.toml --no-cov-on-fail --cov-report term-missing:skip-covered --cov-context=test --cov-report html:/tmp/pipdeptree/.tox/pypy310/tmp/htmlcov --cov-report xml:/tmp/pipdeptree/.tox/coverage.pypy310.xml --junitxml /tmp/pipdeptree/.tox/junit.pypy310.xml tests
============================= test session starts ==============================
platform linux -- Python 3.10.14[pypy-7.3.16-final], pytest-8.2.1, pluggy-1.5.0
cachedir: .tox/pypy310/.pytest_cache
rootdir: /tmp/pipdeptree
configfile: pyproject.toml
plugins: mock-3.14.0, cov-5.0.0, console-scripts-1.4.1
collected 117 items

tests/_models/test_dag.py ............                                   [ 10%]
tests/_models/test_package.py ....................                       [ 27%]
tests/render/test_graphviz.py ...                                        [ 29%]
tests/render/test_json_tree.py ..                                        [ 31%]
tests/render/test_mermaid.py ..                                          [ 33%]
tests/render/test_render.py .....                                        [ 37%]
tests/render/test_text.py ....................                           [ 54%]
tests/test_cli.py .......................                                [ 74%]
tests/test_detect_env.py .....                                           [ 78%]
tests/test_discovery.py .....                                            [ 82%]
tests/test_non_host.py FFF..F                                            [ 88%]
tests/test_pipdeptree.py ..                                              [ 89%]
tests/test_validate.py ...........                                       [ 99%]
tests/test_warning.py .                                                  [100%]

=================================== FAILURES ===================================
________________________ test_custom_interpreter[True] _________________________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-3/test_custom_interpreter_True_0')
mocker = <pytest_mock.plugin.MockerFixture object at 0x00007f7bafbcca68>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x00007f7bafbcd520>
capfd = <_pytest.capture.CaptureFixture object at 0x00007f7bafbcd558>
args_joined = True

    @pytest.mark.parametrize("args_joined", [True, False])
    def test_custom_interpreter(
        tmp_path: Path,
        mocker: MockerFixture,
        monkeypatch: pytest.MonkeyPatch,
        capfd: pytest.CaptureFixture[str],
        args_joined: bool,
    ) -> None:
        # Delete $PYTHONPATH so that it cannot be passed to the custom interpreter process (since we don't know what
        # distribution metadata to expect when it's used).
        monkeypatch.delenv("PYTHONPATH", False)
    
        monkeypatch.chdir(tmp_path)
        result = virtualenv.cli_run([str(tmp_path / "venv"), "--activators", ""])
        py = str(result.creator.exe.relative_to(tmp_path))
        cmd = ["", f"--python={result.creator.exe}"] if args_joined else ["", "--python", py]
        mocker.patch("pipdeptree._discovery.sys.argv", cmd)
        main()
        out, _ = capfd.readouterr()
        found = {i.split("==")[0] for i in out.splitlines()}
        implementation = python_implementation()
        if implementation == "CPython":
            expected = {"pip", "setuptools", "wheel"}
        elif implementation == "PyPy":
            expected = {"cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}  # pragma: no cover
        else:
            raise ValueError(implementation)
        if sys.version_info >= (3, 12):
            expected -= {"setuptools", "wheel"}
>       assert found == expected, out
E       AssertionError: cffi==1.17.0.dev0
E         greenlet==0.4.13
E         hpy==0.9.0
E         └── setuptools [required: >=64.0, installed: 69.5.1]
E         pip==24.0
E         readline==6.2.4.1
E         wheel==0.43.0
E         
E       assert {'cffi', 'gre... 'wheel', ...} == {'cffi', 'gre...ols', 'wheel'}
E         
E         Extra items in the left set:
E         'hpy'
E         '└── setuptools [required: >=64.0, installed: 69.5.1]'
E         Extra items in the right set:
E         'setuptools'
E         Use -v to get more diff

/tmp/pipdeptree/tests/test_non_host.py:48: AssertionError
________________________ test_custom_interpreter[False] ________________________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-3/test_custom_interpreter_False_0')
mocker = <pytest_mock.plugin.MockerFixture object at 0x0000564cf92dd5c8>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x0000564cf92dd558>
capfd = <_pytest.capture.CaptureFixture object at 0x0000564cf92dd4b0>
args_joined = False

    @pytest.mark.parametrize("args_joined", [True, False])
    def test_custom_interpreter(
        tmp_path: Path,
        mocker: MockerFixture,
        monkeypatch: pytest.MonkeyPatch,
        capfd: pytest.CaptureFixture[str],
        args_joined: bool,
    ) -> None:
        # Delete $PYTHONPATH so that it cannot be passed to the custom interpreter process (since we don't know what
        # distribution metadata to expect when it's used).
        monkeypatch.delenv("PYTHONPATH", False)
    
        monkeypatch.chdir(tmp_path)
        result = virtualenv.cli_run([str(tmp_path / "venv"), "--activators", ""])
        py = str(result.creator.exe.relative_to(tmp_path))
        cmd = ["", f"--python={result.creator.exe}"] if args_joined else ["", "--python", py]
        mocker.patch("pipdeptree._discovery.sys.argv", cmd)
        main()
        out, _ = capfd.readouterr()
        found = {i.split("==")[0] for i in out.splitlines()}
        implementation = python_implementation()
        if implementation == "CPython":
            expected = {"pip", "setuptools", "wheel"}
        elif implementation == "PyPy":
            expected = {"cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}  # pragma: no cover
        else:
            raise ValueError(implementation)
        if sys.version_info >= (3, 12):
            expected -= {"setuptools", "wheel"}
>       assert found == expected, out
E       AssertionError: cffi==1.17.0.dev0
E         greenlet==0.4.13
E         hpy==0.9.0
E         └── setuptools [required: >=64.0, installed: 69.5.1]
E         pip==24.0
E         readline==6.2.4.1
E         wheel==0.43.0
E         
E       assert {'cffi', 'gre... 'wheel', ...} == {'cffi', 'gre...ols', 'wheel'}
E         
E         Extra items in the left set:
E         'hpy'
E         '└── setuptools [required: >=64.0, installed: 69.5.1]'
E         Extra items in the right set:
E         'setuptools'
E         Use -v to get more diff

/tmp/pipdeptree/tests/test_non_host.py:48: AssertionError
___________________ test_custom_interpreter_with_local_only ____________________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-3/test_custom_interpreter_with_l0')
mocker = <pytest_mock.plugin.MockerFixture object at 0x0000564cfacb9c58>
capfd = <_pytest.capture.CaptureFixture object at 0x0000564cfacb9bb0>

    def test_custom_interpreter_with_local_only(
        tmp_path: Path,
        mocker: MockerFixture,
        capfd: pytest.CaptureFixture[str],
    ) -> None:
        venv_path = str(tmp_path / "venv")
    
        result = virtualenv.cli_run([venv_path, "--system-site-packages", "--activators", ""])
    
        cmd = ["", f"--python={result.creator.exe}", "--local-only"]
        mocker.patch("pipdeptree._discovery.sys.prefix", venv_path)
        mocker.patch("pipdeptree._discovery.sys.argv", cmd)
        main()
        out, _ = capfd.readouterr()
        found = {i.split("==")[0] for i in out.splitlines()}
        implementation = python_implementation()
        if implementation == "CPython":
            expected = {"pip", "setuptools", "wheel"}
        elif implementation == "PyPy":  # pragma: no cover
            expected = {"cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}  # pragma: no cover
        else:
            raise ValueError(implementation)  # pragma: no cover
        if sys.version_info >= (3, 12):
            expected -= {"setuptools", "wheel"}  # pragma: no cover
>       assert found == expected, out
E       AssertionError: pip==24.0
E         setuptools==69.5.1
E         wheel==0.43.0
E         
E       assert {'pip', 'setuptools', 'wheel'} == {'cffi', 'gre...ols', 'wheel'}
E         
E         Extra items in the right set:
E         'cffi'
E         'greenlet'
E         'readline'
E         Use -v to get more diff

tests/test_non_host.py:75: AssertionError
__________ test_custom_interpreter_ensure_pythonpath_envar_is_honored __________

tmp_path = PosixPath('/tmp/pytest-of-mgorny/pytest-3/test_custom_interpreter_ensure0')
mocker = <pytest_mock.plugin.MockerFixture object at 0x0000564cf9fb3980>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x0000564cf9fb3910>
capfd = <_pytest.capture.CaptureFixture object at 0x0000564cf9fb3830>

    def test_custom_interpreter_ensure_pythonpath_envar_is_honored(
        tmp_path: Path,
        mocker: MockerFixture,
        monkeypatch: pytest.MonkeyPatch,
        capfd: pytest.CaptureFixture[str],
    ) -> None:
        # ensures that we honor $PYTHONPATH when passing it to the custom interpreter process
        venv_path = str(tmp_path / "venv")
        result = virtualenv.cli_run([venv_path, "--activators", ""])
    
        another_path = tmp_path / "another-path"
        fake_dist = another_path / "foo-1.2.3.dist-info"
        fake_dist.mkdir(parents=True)
        fake_metadata = fake_dist / "METADATA"
        with fake_metadata.open("w") as f:
            f.write("Metadata-Version: 2.3\n" "Name: foo\n" "Version: 1.2.3\n")
        cmd = ["", f"--python={result.creator.exe}"]
        mocker.patch("pipdeptree._discovery.sys.argv", cmd)
        monkeypatch.setenv("PYTHONPATH", str(another_path))
        main()
        out, _ = capfd.readouterr()
        found = {i.split("==")[0] for i in out.splitlines()}
        implementation = python_implementation()
        if implementation == "CPython":
            expected = {"foo", "pip", "setuptools", "wheel"}
        elif implementation == "PyPy":  # pragma: cpython no cover
            expected = {"foo", "cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}
        else:  # pragma: no cover
            raise ValueError(implementation)
        if sys.version_info >= (3, 12):  # pragma: >=3.12 cover
            expected -= {"setuptools", "wheel"}
>       assert found == expected, out
E       AssertionError: cffi==1.17.0.dev0
E         foo==1.2.3
E         greenlet==0.4.13
E         hpy==0.9.0
E         └── setuptools [required: >=64.0, installed: 69.5.1]
E         pip==24.0
E         readline==6.2.4.1
E         wheel==0.43.0
E         
E       assert {'cffi', 'foo...eadline', ...} == {'cffi', 'foo...uptools', ...}
E         
E         Extra items in the left set:
E         'hpy'
E         '└── setuptools [required: >=64.0, installed: 69.5.1]'
E         Extra items in the right set:
E         'setuptools'
E         Use -v to get more diff

tests/test_non_host.py:164: AssertionError
=============================== warnings summary ===============================
tests/test_pipdeptree.py::test_console[inprocess]
  /tmp/pipdeptree/tests/test_pipdeptree.py:16: DeprecationWarning: script_runner commands should be passed as a single sequence, not as multiple arguments.
  Replace `script_runner.run(a, b, c)` calls with `script_runner.run([a, b, c])`
    result = script_runner.run("pipdeptree", "--help")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---------- generated xml file: /tmp/pipdeptree/.tox/junit.pypy310.xml ----------
=========================== short test summary info ============================
FAILED tests/test_non_host.py::test_custom_interpreter[True] - AssertionError...
FAILED tests/test_non_host.py::test_custom_interpreter[False] - AssertionErro...
FAILED tests/test_non_host.py::test_custom_interpreter_with_local_only - Asse...
FAILED tests/test_non_host.py::test_custom_interpreter_ensure_pythonpath_envar_is_honored
================== 4 failed, 113 passed, 1 warning in 12.61s ===================
pypy310: exit 1 (14.30 seconds) /tmp/pipdeptree> python -m pytest --cov /tmp/pipdeptree/.tox/pypy310/lib/pypy3.10/site-packages/pipdeptree --cov /tmp/pipdeptree/tests --cov-config=pyproject.toml --no-cov-on-fail --cov-report term-missing:skip-covered --cov-context=test --cov-report html:/tmp/pipdeptree/.tox/pypy310/tmp/htmlcov --cov-report xml:/tmp/pipdeptree/.tox/coverage.pypy310.xml --junitxml /tmp/pipdeptree/.tox/junit.pypy310.xml tests pid=342108
.pkg: _exit> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True hatchling.build
  pypy310: FAIL code 1 (15.34=setup[1.04]+cmd[14.30] seconds)
  evaluation failed :( (15.41 seconds)
```

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -24,6 +24,8 @@ jobs:
           - "3.10"
           - "3.9"
           - "3.8"
+          - "pypy3.10"
+          - "pypy3.9"
     steps:
       - name: Install OS dependencies
         run: |
@@ -47,9 +49,14 @@ jobs:
           python-version: ${{ matrix.py }}
       - name: Pick environment to run
         run: |
-          import codecs; import os; import sys
-          env = "TOXENV=py3{}\n".format(sys.version_info[1])
-          print("Picked:\n{}for{}".format(env, sys.version))
+          import codecs; import os; import sys; import platform
+          impl = platform.python_implementation()
+          if impl == "CPython":
+            env = "TOXENV=py3"
+          else:
+            env = "TOXENV=pypy3"
+          env += "{}\n".format(sys.version_info.minor)
+          print("Picked:\n{}for {}".format(env, sys.version))
           with codecs.open(os.environ["GITHUB_ENV"], "a", "utf-8") as file_handler:
                file_handler.write(env)
         shell: python

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -51,10 +51,10 @@ jobs:
         run: |
           import codecs; import os; import sys; import platform
           impl = platform.python_implementation()
-          if impl == "CPython":
-            env = "TOXENV=py3"
-          else:
+          if impl == "PyPy":
             env = "TOXENV=pypy3"
+          else:
+            env = "TOXENV=py3"
           env += "{}\n".format(sys.version_info.minor)
           print("Picked:\n{}for {}".format(env, sys.version))
           with codecs.open(os.environ["GITHUB_ENV"], "a", "utf-8") as file_handler:

--- a/tests/test_non_host.py
+++ b/tests/test_non_host.py
@@ -48,6 +48,7 @@ def test_custom_interpreter(
     result = virtualenv.cli_run([str(tmp_path / "venv"), "--activators", ""])
     py = str(result.creator.exe.relative_to(tmp_path))
     cmd = ["", f"--python={result.creator.exe}"] if args_joined else ["", "--python", py]
+    cmd += ["--all", "--depth", "0"]
     mocker.patch("pipdeptree._discovery.sys.argv", cmd)
     main()
     out, _ = capfd.readouterr()
@@ -84,7 +85,7 @@ def test_custom_interpreter_with_user_only(
     venv_path = str(tmp_path / "venv")
     result = virtualenv.cli_run([venv_path, "--activators", ""])
 
-    cmd = ["", f"--python={result.creator.exe}", "--user-only"]
+    cmd = ["", f"--python={result.creator.exe}", "--all", "--depth", "0", "--user-only"]
     mocker.patch("pipdeptree.__main__.sys.argv", cmd)
     main()
     out, err = capfd.readouterr()
@@ -147,10 +148,10 @@ def test_custom_interpreter_ensure_pythonpath_envar_is_honored(
     fake_metadata = fake_dist / "METADATA"
     with fake_metadata.open("w") as f:
         f.write("Metadata-Version: 2.3\n" "Name: foo\n" "Version: 1.2.3\n")
-    cmd = ["", f"--python={result.creator.exe}"]
+    cmd = ["", f"--python={result.creator.exe}", "--all", "--depth", "0"]
     mocker.patch("pipdeptree._discovery.sys.argv", cmd)
     monkeypatch.setenv("PYTHONPATH", str(another_path))
     main()
     out, _ = capfd.readouterr()
     found = {i.split("==")[0] for i in out.splitlines()}
-    assert {*expected_venv_pkgs, "foo"} == found
+    assert {*expected_venv_pkgs, "foo"} == found, out

--- a/tests/test_non_host.py
+++ b/tests/test_non_host.py
@@ -16,6 +16,21 @@ if TYPE_CHECKING:
     from pytest_mock import MockerFixture
 
 
+@pytest.fixture(scope="session")
+def expected_venv_pkgs() -> frozenset[str]:
+    implementation = python_implementation()
+    if implementation == "CPython":  # pragma: cpython cover
+        expected = {"pip", "setuptools", "wheel"}
+    elif implementation == "PyPy":  # pragma: pypy cover
+        expected = {"cffi", "greenlet", "pip", "readline", "hpy", "setuptools", "wheel"}
+    else:  # pragma: no cover
+        raise ValueError(implementation)
+    if sys.version_info >= (3, 12):  # pragma: >=3.12 cover
+        expected -= {"setuptools", "wheel"}
+
+    return frozenset(expected)
+
+
 @pytest.mark.parametrize("args_joined", [True, False])
 def test_custom_interpreter(
     tmp_path: Path,
@@ -23,6 +38,7 @@ def test_custom_interpreter(
     monkeypatch: pytest.MonkeyPatch,
     capfd: pytest.CaptureFixture[str],
     args_joined: bool,
+    expected_venv_pkgs: frozenset[str],
 ) -> None:
     # Delete $PYTHONPATH so that it cannot be passed to the custom interpreter process (since we don't know what
     # distribution metadata to expect when it's used).
@@ -36,16 +52,8 @@ def test_custom_interpreter(
     main()
     out, _ = capfd.readouterr()
     found = {i.split("==")[0] for i in out.splitlines()}
-    implementation = python_implementation()
-    if implementation == "CPython":
-        expected = {"pip", "setuptools", "wheel"}
-    elif implementation == "PyPy":
-        expected = {"cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}  # pragma: no cover
-    else:
-        raise ValueError(implementation)
-    if sys.version_info >= (3, 12):
-        expected -= {"setuptools", "wheel"}
-    assert found == expected, out
+
+    assert expected_venv_pkgs == found, out
 
 
 def test_custom_interpreter_with_local_only(
@@ -54,7 +62,6 @@ def test_custom_interpreter_with_local_only(
     capfd: pytest.CaptureFixture[str],
 ) -> None:
     venv_path = str(tmp_path / "venv")
-
     result = virtualenv.cli_run([venv_path, "--system-site-packages", "--activators", ""])
 
     cmd = ["", f"--python={result.creator.exe}", "--local-only"]
@@ -63,16 +70,10 @@ def test_custom_interpreter_with_local_only(
     main()
     out, _ = capfd.readouterr()
     found = {i.split("==")[0] for i in out.splitlines()}
-    implementation = python_implementation()
-    if implementation == "CPython":
-        expected = {"pip", "setuptools", "wheel"}
-    elif implementation == "PyPy":  # pragma: no cover
-        expected = {"cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}  # pragma: no cover
-    else:
-        raise ValueError(implementation)  # pragma: no cover
-    if sys.version_info >= (3, 12):
-        expected -= {"setuptools", "wheel"}  # pragma: no cover
-    assert found == expected, out
+    expected = {"pip", "setuptools", "wheel"}
+    if sys.version_info >= (3, 12):  # pragma: >=3.12 cover
+        expected -= {"setuptools", "wheel"}
+    assert expected == found, out
 
 
 def test_custom_interpreter_with_user_only(
@@ -81,7 +82,6 @@ def test_custom_interpreter_with_user_only(
     # ensures there is no output when --user-only and --python are passed
 
     venv_path = str(tmp_path / "venv")
-
     result = virtualenv.cli_run([venv_path, "--activators", ""])
 
     cmd = ["", f"--python={result.creator.exe}", "--user-only"]
@@ -127,7 +127,7 @@ def test_custom_interpreter_with_user_only_and_system_site_pkgs_enabled(
     found = {i.split("==")[0] for i in out.splitlines()}
     expected = {"bar"}
 
-    assert found == expected
+    assert expected == found
 
 
 def test_custom_interpreter_ensure_pythonpath_envar_is_honored(
@@ -135,6 +135,7 @@ def test_custom_interpreter_ensure_pythonpath_envar_is_honored(
     mocker: MockerFixture,
     monkeypatch: pytest.MonkeyPatch,
     capfd: pytest.CaptureFixture[str],
+    expected_venv_pkgs: frozenset[str],
 ) -> None:
     # ensures that we honor $PYTHONPATH when passing it to the custom interpreter process
     venv_path = str(tmp_path / "venv")
@@ -152,13 +153,4 @@ def test_custom_interpreter_ensure_pythonpath_envar_is_honored(
     main()
     out, _ = capfd.readouterr()
     found = {i.split("==")[0] for i in out.splitlines()}
-    implementation = python_implementation()
-    if implementation == "CPython":
-        expected = {"foo", "pip", "setuptools", "wheel"}
-    elif implementation == "PyPy":  # pragma: cpython no cover
-        expected = {"foo", "cffi", "greenlet", "pip", "readline", "setuptools", "wheel"}
-    else:  # pragma: no cover
-        raise ValueError(implementation)
-    if sys.version_info >= (3, 12):  # pragma: >=3.12 cover
-        expected -= {"setuptools", "wheel"}
-    assert found == expected, out
+    assert {*expected_venv_pkgs, "foo"} == found

--- a/tests/test_non_host.py
+++ b/tests/test_non_host.py
@@ -85,7 +85,7 @@ def test_custom_interpreter_with_user_only(
     venv_path = str(tmp_path / "venv")
     result = virtualenv.cli_run([venv_path, "--activators", ""])
 
-    cmd = ["", f"--python={result.creator.exe}", "--all", "--depth", "0", "--user-only"]
+    cmd = ["", f"--python={result.creator.exe}", "--user-only"]
     mocker.patch("pipdeptree.__main__.sys.argv", cmd)
     main()
     out, err = capfd.readouterr()

